### DISCUSSION TITLE
Guide for new contributors, setting up IDE?

### DISCUSSION THREAD

**Discussion by Alatius** [fontforge/fontforge#5083] [Q&A] [answered]
Hi! Is there a guide somewhere, or general advice on how to set up a development environment to best work with the FontForge code? What IDE are you using when coding on FontForge? I just want to be able to do the basic stuff to begin with, like setting breakpoints and inspecting variable values. Personally I am mainly interested in solutions for Linux or Mac, but I am sure advice for Windows will be welcome to many prospective contributors!

**Comment by ctrlcctrlv** [comment, marked as answer]
I don't think so; I just use `vim` to patch FontForge, I think @skef does the same. To make tags, `:!ctags -R .`. To build, `:!(cd build; ninja)` To run, `:!(cd build; bin/fontforge)`. To build and run only on successful build with debugging, `:!(cd build; ninja && gdb -ex 'run' bin/fontforge)`. Make sure to configure CMake to compile a developer build (with a symbol table and `-O0` if you will do this, see wiki).

**Reply by Alatius** [reply]
I see! The tags feature sounds nice (I haven't used vim enough to have tested that). But what you describe does not sound very convenient for debugging, is it? Like, for example, the reason why I asked is that I think I have found a bug somewhere in the spline simplification function, and to figure that out, it would be immensely helpful to place breakpoints, step through the algorithm, and see the variables change as I go. Hence my question.

**Reply by ctrlcctrlv** [reply]
@Alatius If I want to do that I just change the final command from `bin/fontforge` to `gdb -ex 'run' bin/fontforge` and then do my work in `gdb`. So I fail to understand how it's inconvenient.

**Reply by Alatius** [reply]
Not inconvenient at all, I am sure, if you are used to working with gdb directly (which I am not, at all). I will give it a try.

**Reply by ctrlcctrlv** [reply]
Yes, sorry I can't be of more help in setting up an IDE; I don't use them as I find them too unpredictable and limiting, especially for FontForge development. Another thing to keep in mind is FontForge's highly unorthodox indentation style (mixed tabs/spaces) which requires specific `vim` trickery on my part to make sure I don't screw up in my patches……

**Reply by Alatius** [reply]
Well, I must say that I vastly over-estimated the difficulty of using GDB, so thank you for pushing me to use it! :-) (The findings from my debugging are summarized in issue #5093.)

**Reply by ctrlcctrlv** [reply]
Glad to hear. By the way @Alatius it is also possible to use `lldb` if you need to due to some difficult to find crash. I have done it before. I assume that it would be equally possible to use a GUI frontend to GDB (so just a plain debugger, not an editor-integrated one).

In `neovim` (which I alias to `vim`) I just `:vsplit|te` and run the `gdb` in there. I know it takes a bit of getting used to but I think most long-time FOSS developers prefer working this way. Been bitten by too many IDE-causing yak-shaving marathons.

**Comment by Alatius** [comment]
I am curious to hear from other contributors, how does/did your FontForge development environment look? Ping @jtanx @monkeyiq @JoesCat @frank-trampe @khaledhosny @gioele @rrthomas @davelab6 @linusromer @dscorbett

**Reply by ctrlcctrlv** [reply]
(Although I just marked my answer as the answer I'm very interested if any active developer goes against my way of working on FontForge, lol. Not trying to fanboy&dagger; one particular way, just in my discussions with other devs my way seems to be standard. Although I think @JoesCat uses Emacs, but the broad strokes remain the same as that's a text editor too.)

<sub>&dagger; Fanfurry?</sub>

**Reply by dscorbett** [reply]
I use vim for FontForge development. I don’t do anything complex enough to require a debugger.
